File: PopularShows/app/src/main/java/com/lgcode/popularshows/adapters/TVShowsAdapter.java
package com.lgcode.popularshows.adapters;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.otto.Bus;
import com.squareup.picasso.Picasso;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import glb.lg.popularshows.R;
import com.lgcode.popularshows.TvShow;

/**
 * Created by leojg on 11/9/16.
 */
public class TVShowsAdapter extends RecyclerView.Adapter<TVShowsAdapter.TVShowsViewHolder> {

    private List<TvShow> tvShowItems;
    private Bus bus;
    private Context context;
    private int lastPosition = -1;

    public TVShowsAdapter(List<TvShow> tvShowItems, Context context, Bus bus) {
        this.tvShowItems = tvShowItems;
        this.bus = bus;
        this.context = context;
    }

    @Override
    public TVShowsViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.tv_show_item, parent, false);
        return new TVShowsViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final TVShowsViewHolder viewHolder, int position) {

        final TvShow tvShow = tvShowItems.get(position);
        viewHolder.tvShowTitle.setText(tvShow.getTitle());
        viewHolder.tvShowVotes.setText(String.valueOf(tvShow.getVoteAverage()));
        Picasso.with(context).load(tvShow.getImage()).into(viewHolder.tvShowThumbnail);
        viewHolder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                bus.post(new TVShowsGetDetailViewEvent(tvShow.getId()));
            }
        });
        setAnimation(viewHolder.itemView, position);
    }

    private void setAnimation(View view, int position) {
        if (position > lastPosition) {
            view.animate().cancel();
            view.setTranslationY(100);
            view.setAlpha(0);
            view.animate().alpha(1.0f).translationY(0).setDuration(200).setStartDelay(100);
            lastPosition = position;
        }
    }

    @Override
    public int getItemCount() {
        return (null != tvShowItems ? tvShowItems.size() : 0);
    }

    class TVShowsViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.tv_show_thumbnail) ImageView tvShowThumbnail;
        @BindView(R.id.tv_show_title) TextView tvShowTitle;
        @BindView(R.id.tv_show_votes) TextView tvShowVotes;

        TVShowsViewHolder(View view) {
            super(view);
            ButterKnife.setDebug(true);
            ButterKnife.bind(this, view);
        }
    }

    public class TVShowsGetDetailViewEvent {
        public final Integer showId;

        TVShowsGetDetailViewEvent(Integer showId) {
            this.showId = showId;
        }
    }
}

File: KtPopularShows/app/src/main/java/example/lgcode/ktpopularshows/adapter/TVShowsAdapter.kt
package example.lgcode.ktpopularshows.adapter

import android.content.Context
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.squareup.picasso.Picasso
import example.lgcode.ktpopularshows.BuildConfig
import example.lgcode.ktpopularshows.R
import example.lgcode.ktpopularshows.domain.TVShow
import example.lgcode.ktpopularshows.domain.TVShowDetailEvent
import io.reactivex.observers.DisposableObserver
import kotlinx.android.synthetic.main.tv_show_item.view.*

class TVShowsAdapter(
        private val context: Context,
        private val observer: DisposableObserver<TVShowDetailEvent>
): RecyclerView.Adapter<TVShowsAdapter.TVShowsViewHolder>() {

    private val tvShowItems: MutableList<TVShow> = mutableListOf()
    private var lastPosition = -1
    private lateinit var recyclerView: RecyclerView

    fun addAll(tvShows: List<TVShow>) {
        tvShowItems.addAll(tvShows)
    }

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): TVShowsViewHolder {
        if (isVertical()) {
            return TVShowsViewHolder(LayoutInflater.from(parent!!.context).inflate(R.layout.tv_show_item, parent, false))
        } else {
            return TVShowsViewHolder(LayoutInflater.from(parent!!.context).inflate(R.layout.similar_tv_show_item, parent, false))

        }
    }

    override fun onBindViewHolder(holder: TVShowsViewHolder?, position: Int) {
        val tvShow = tvShowItems.get(position)

        holder!!.view.tv_show_title.text = tvShow.title
        holder.view.tv_show_votes.text = tvShow.voteAverage.toString()
        Picasso.with(context)
                .load(StringBuilder(BuildConfig.SMALL_IMAGE_BASE_URL).append(tvShow.image).toString())
                .into(holder.view.tv_show_thumbnail)
        holder.view.setOnClickListener {
            observer.onNext(TVShowDetailEvent(tvShow))
        }

        setAnimation(holder.view, position)
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView?) {
        super.onAttachedToRecyclerView(recyclerView)
        this.recyclerView = recyclerView!!
    }

    private fun setAnimation(view: View, position: Int) {
        if (position > lastPosition) {
            view.animate().cancel()
            view.translationY = 100f
            view.alpha = 0f
            if (isVertical()) {
                view.animate().alpha(1.0f).translationY(0f).setDuration(200).startDelay = 100
            } else {
                view.animate().alpha(1.0f).translationX(0f).setDuration(200).startDelay = 100
            }
            lastPosition = position
        }
    }

    private fun isVertical(): Boolean {
        return (recyclerView.layoutManager as LinearLayoutManager).orientation == LinearLayoutManager.VERTICAL
    }

    override fun getItemCount(): Int {
        return tvShowItems.size
    }

    class TVShowsViewHolder(val view: View): RecyclerView.ViewHolder(view)

}